File: backend-go/internal/handler/character.go
package handler

import (
	"strconv"

	"github.com/CedrieGuan/huobao-drama-go/backend-go/internal/database"
	"github.com/CedrieGuan/huobao-drama-go/backend-go/internal/util"
	"github.com/gin-gonic/gin"
)

// ---------------------------------------------------------------------------
// PUT /api/v1/characters/:id — Update character
// ---------------------------------------------------------------------------

// UpdateCharacter updates select fields of a character.
// Body: { name?, role?, description?, appearance?, personality?,
//
//	voice_style?, voice_provider?, image_url?, local_path?,
//	reference_images?, seed_value?, sort_order? }
//
// Accepts both snake_case and camelCase keys (matching TS behavior).
// If voice_style is changed, voice_sample_url is reset to null.
func (h *Handler) UpdateCharacter(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		util.BadRequest(c, "invalid id")
		return
	}

	var body map[string]interface{}
	if err := c.ShouldBindJSON(&body); err != nil {
		util.BadRequest(c, "invalid request body")
		return
	}

	updates := map[string]interface{}{
		"updated_at": database.Now(),
	}

	// Map of allowed JSON keys to database column names.
	// The TS backend accepts both snake_case and camelCase, so we handle both.
	allowedFields := map[string]string{
		// snake_case keys (preferred)
		"name":             "name",
		"role":             "role",
		"description":      "description",
		"appearance":       "appearance",
		"personality":      "personality",
		"voice_style":      "voice_style",
		"voice_provider":   "voice_provider",
		"image_url":        "image_url",
		"local_path":       "local_path",
		"reference_images": "reference_images",
		"seed_value":       "seed_value",
		"sort_order":       "sort_order",
		"voice_sample_url": "voice_sample_url",
		"thumbnail":        "thumbnail",
	}

	for jsonKey, dbCol := range allowedFields {
		if v, ok := body[jsonKey]; ok {
			updates[dbCol] = v
		}
	}

	// If voice_style is being changed, reset voice_sample_url to null.
	// This matches the TS behavior: changing the voice invalidates the old sample.
	_, hasVoiceStyle := body["voice_style"]
	if hasVoiceStyle {
		updates["voice_sample_url"] = nil
	}

	h.DB.Model(&database.Character{}).Where("id = ?", id).Updates(updates)
	util.Success(c, nil)
}

// ---------------------------------------------------------------------------
// DELETE /api/v1/characters/:id — Soft delete character
// ---------------------------------------------------------------------------

// DeleteCharacter soft-deletes a character by setting deleted_at.
// This is a soft delete (matching the TypeScript backend behavior).
func (h *Handler) DeleteCharacter(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		util.BadRequest(c, "invalid id")
		return
	}

	h.DB.Model(&database.Character{}).Where("id = ?", id).
		Update("deleted_at", database.Now())

	util.Success(c, nil)
}

File: macapp/HuobaoDrama/HuobaoDrama/Models/Character.swift
import Foundation

struct Character: Codable, Identifiable, Sendable, Equatable {
    let id: Int
    var dramaId: Int
    var name: String
    var role: String?
    var description: String?
    var appearance: String?
    var personality: String?
    var voiceStyle: String?
    var imageUrl: String?
    var referenceImages: [String]
    var seedValue: String?
    var sortOrder: Int?
    var localPath: String?
    var voiceSampleUrl: String?
    var voiceProvider: String?
    var createdAt: String
    var updatedAt: String

    enum CodingKeys: String, CodingKey {
        case id, name, role, description, appearance, personality
        case dramaId = "drama_id"
        case voiceStyle = "voice_style"
        case imageUrl = "image_url"
        case referenceImages = "reference_images"
        case seedValue = "seed_value"
        case sortOrder = "sort_order"
        case localPath = "local_path"
        case voiceSampleUrl = "voice_sample_url"
        case voiceProvider = "voice_provider"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(Int.self, forKey: .id)
        dramaId = try c.decodeIfPresent(Int.self, forKey: .dramaId) ?? 0
        name = try c.decode(String.self, forKey: .name)
        role = try c.decodeIfPresent(String.self, forKey: .role)
        description = try c.decodeIfPresent(String.self, forKey: .description)
        appearance = try c.decodeIfPresent(String.self, forKey: .appearance)
        personality = try c.decodeIfPresent(String.self, forKey: .personality)
        voiceStyle = try c.decodeIfPresent(String.self, forKey: .voiceStyle)
        imageUrl = try c.decodeIfPresent(String.self, forKey: .imageUrl)
        seedValue = try c.decodeIfPresent(String.self, forKey: .seedValue)
        sortOrder = try c.decodeIfPresent(Int.self, forKey: .sortOrder)
        localPath = try c.decodeIfPresent(String.self, forKey: .localPath)
        voiceSampleUrl = try c.decodeIfPresent(String.self, forKey: .voiceSampleUrl)
        voiceProvider = try c.decodeIfPresent(String.self, forKey: .voiceProvider)
        createdAt = try c.decode(String.self, forKey: .createdAt)
        updatedAt = try c.decode(String.self, forKey: .updatedAt)
        if let raw = try c.decodeIfPresent(String.self, forKey: .referenceImages),
           let arr = try? JSONDecoder().decode([String].self, from: Data(raw.utf8)) {
            referenceImages = arr
        } else {
            referenceImages = []
        }
    }
}
